src/scheduler/Executor.cpp

```diff
@@ -99,6 +116,189 @@ void Executor::finish()
     deadThreads.clear();
 }
 
+std::vector<std::pair<uint32_t, int32_t>> Executor::executeThreads(
+  std::shared_ptr<faabric::BatchExecuteRequest> req,
+  const std::vector<faabric::util::SnapshotMergeRegion>& mergeRegions)
+{
+    SPDLOG_DEBUG("Executor {} executing {} threads", id, req->messages_size());
+
+    faabric::Message& msg = req->mutable_messages()->at(0);
+    std::string snapshotKey = faabric::util::getMainThreadSnapshotKey(msg);
+    std::string funcStr = faabric::util::funcToString(msg, false);
+
+    std::shared_ptr<faabric::util::SnapshotData> snap = nullptr;
+    bool exists = false;
+    {
+        faabric::util::SharedLock lock(threadExecutionMutex);
+        exists = reg.snapshotExists(snapshotKey);
+    }
+
+    if (!exists) {
+        faabric::util::FullLock lock(threadExecutionMutex);
+        if (!reg.snapshotExists(snapshotKey)) {
+            SPDLOG_DEBUG(
+              "Creating main thread snapshot: {} for {}", snapshotKey, funcStr);
+
+            snap =
+              std::make_shared<faabric::util::SnapshotData>(getMemoryView());
+            reg.registerSnapshot(snapshotKey, snap);
+        } else {
+            exists = true;
+        }
+    }
+
+    if (exists) {
+        SPDLOG_DEBUG(
+          "Main thread snapshot exists: {} for {}", snapshotKey, funcStr);
+
+        // Get main snapshot
+        snap = reg.getSnapshot(snapshotKey);
+        std::span<uint8_t> memView = getMemoryView();
+
+        // Get dirty regions since last batch of threads
+        tracker.stopTracking(memView);
+        tracker.stopThreadLocalTracking(memView);
+
+        std::vector<std::pair<uint32_t, uint32_t>> dirtyRegions =
+          tracker.getBothDirtyOffsets(memView);
+
+        // Apply changes to snapshot
+        snap->fillGapsWithOverwriteRegions();
+        std::vector<faabric::util::SnapshotDiff> updates =
+          snap->diffWithDirtyRegions(memView, dirtyRegions);
+
+        if (updates.empty()) {
+            SPDLOG_TRACE(
+              "No updates to main thread snapshot for {} from {} dirty regions",
+              faabric::util::funcToString(msg, false),
+              dirtyRegions.size());
+        } else {
+            SPDLOG_DEBUG("Updating main thread snapshot for {} with {} diffs",
+                         faabric::util::funcToString(msg, false),
+                         updates.size());
+            snap->queueDiffs(updates);
+            snap->writeQueuedDiffs();
+        }
+
+        snap->clearMergeRegions();
+    }
+
+    // Now we have to apply the merge regions for this parallel section
+    for (const auto& mr : mergeRegions) {
+        snap->addMergeRegion(
+          mr.offset, mr.length, mr.dataType, mr.operation, true);
+    }
+
+    // TODO - here the main thread will wait, so technically frees up a slot
+    // that could be used.
+    std::string cacheKey =
+      std::to_string(msg.appid()) + "_" + std::to_string(req->messages_size());
+    bool hasCachedDecision = false;
```

Comment:
Could we make this a `std::atomic<bool>`?

Reply:
This (and the other variables you mentioned) is a stack-local, not shared across threads so atomics don't really make sense here. You only need atomic accesses if the same variable (in the same memory location) is accessed from multiple threads without being in a mutex critical section.